make duplicate def an error, fix any resulting issues with the legacy registration API

bdhirsh opened the pull request:
Make duplicate definitions in the Dispatcher registration API an error.

Context: The dispatcher has a new API as well as a more legacy API. One advantage of the new API is that it only requires you to specify the full schema of the operator that you're registering in one place (where it is "defined"). All other registrations to different kernels only need the op name and the corresponding kernel. Logically, we'd like to enforce that a user only def()'s an operator once, as this lets us clean up a lot of the AliasAnalysis logic used to deal with duplicate definitions.

The legacy API makes it relatively easier to def() an operator multiple times, and we actually do so in several places in source. Namely:


1. C10_EXPORT_CAFFE2_OP_TO_C10 macros. These macros are used for caffe2 ops, and the macro they use for registration with the dispatcher uses the legacy API in a way that results in duplicate def() calls.

2. Tests. We have lots of tests of the legacy registration API. I tried to clean them up in a sane way- a combination of directly fixing, migrating to the new API, or deleting the tests.

**What this PR does**
1. Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This also required creating two new macros (TORCH_LIBRARY_FRAGMENT_UNIQUE and TORCH_LIBRARY_IMPL_UNIQUE). They are identical to their counterparts, but avoids any variable name collisions when you use each macro multiple times with the same namespace/dispatch key in the same translation unit. This is the case for the caffe2 macros.

2. Updated tests

Only tests using the legacy registration API caused failures. I took all the failed tests and (for the most part) did the following):

(a) Fixed it, if possible, if the test semantically tested the same thing (still using the legacy API)
(b) migrated it to use the new registration API (I did this in tests for which I couldn't figure out a way to get the legacy API to correctly express the test's intent without causing a duplicate def()).
(c) Deleted it (I only did this for tests in op_registration_test. Mostly tests that seemed to be testing the semantic behavior of using duplicate def()'s).

3. Renaming `TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY` to `TORCH_LIBRARY_FRAGMENT`.

Doing this because I'm also planning on migrating custom fb ops to use the new API. Some will require the above macro, since many ops in different files are dumped into the same namespace.

4. Deleting caffe2 operators under the namespace `__caffe2` (NOTE: BC-Breaking).

I did this for two reasons:
(a) I don't see them used anywhere in fbcode, from a grep search. This applies to 5 ops. Each is registered twice (the same operator), under both the `_caffe2` and `__caffe2` namespaces.
(b) Doing so allowed me to hardcode the `_caffe2` namespace into the EXPORT... macro, which I think is preferable to updating all call sites of the macro to pass in a custom namespace. This macro is only meant to be used for caffe2 ops.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46923 make duplicate def an error, fix any resulting issues with the legacy registration API**

Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

bdhirsh committed:
[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

ezyang commented on torch/library.h:
```diff
@@ -650,6 +650,17 @@ class TorchLibraryInit final {
   void TORCH_LIBRARY_init_ ## ns (torch::Library& m)
 
+
+// TODO: comment, maybe fold this into TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY
```
Yes, please do

ezyang commented on torch/library.h:
```diff
@@ -717,6 +728,20 @@ class TorchLibraryInit final {
   void TORCH_LIBRARY_IMPL_init_ ## ns ## _ ## k (torch::Library& m)
 
+// TODO: meaningful comment
+#define TORCH_LIBRARY_IMPL_FRAGMENT(ns, k, m, uid) \
```
FRAGMENT is not needed for IMPL, as IMPL already allows multiple blocks for same namespace

bdhirsh replied:
yes- the issue with using IMPL directly is that it defines a bunch of static variables and gives them names uniquely identified by their namespace. This is a problem for the macro, since it gets called multiple times for different operators, all of which go into the _caffe2:: namespace.

The new macro I added isn't particularly elegant, but it resolves that issue by letting you pass in a unique identifier to avoid name collisions- in this case I'm passing in the particular OperatorName that the macro is registering.

Let me know if that makes sense (and also if you have a cleaner solution- I found usage of a __COUNTER__ macro that looked promising but I wasn't able to get it to work (maybe it's platform dependent).

I also might be able to fold into/replace TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY with TORCH_LIBRARY_FRAGMENT. I just want to dig into how that macro is used a little more first.

ezyang commented on aten/src/ATen/core/boxing/impl/kernel_stackbased_test.cpp:
```diff
@@ -72,10 +72,10 @@ TEST(OperatorRegistrationTest_StackBasedKernel, givenKernel_whenRegistered_thenC
+      .op("_test::my_op(Tensor dummy, int input) -> int", RegisterOperators::options().kernel<&incrementKernel>(DispatchKey::CPU)
+                                                                                      .kernel<&errorKernel>(DispatchKey::CUDA))
+      .op("_test::error(Tensor dummy, int input) -> int", RegisterOperators::options().kernel<&errorKernel>(DispatchKey::CPU)
+                                                                                      .kernel<&errorKernel>(DispatchKey::CUDA));
```
Heh.

bdhirsh commented on caffe2/core/export_caffe2_op_to_c10.h:
```diff
@@ -191,48 +194,52 @@ inline FunctionSchema make_function_schema_for_c10(const char* schema_str) {
-    OperatorName, OperatorSchema, OperatorClass)                        \
-  C10_EXPORT_CAFFE2_OP_TO_C10_SCHEMA_ONLY(OperatorName, OperatorSchema) \
+    TORCH_LIBRARY_IMPL_FRAGMENT(_caffe2, CPU, m, OperatorName) {             \
+        m.impl("_caffe2::" #OperatorName,                                    \
```
Just want to call out that I'm hardcoding the namespace (_caffe2::) directly in the macro, since passing it in would require modifying all of the call-sites. I think this is fine, but there are a handful of caffe2 ops that actually don't obey this convention- they register an op twice, once under `_caffe2` and a second time under `__caffe2`. A grep/fbgs search told me that the ops in the second namespace aren't actually used anywhere, so for now I'm deleting them (see `caffe2/operators/bbox_transform_op.cc`)

ezyang replied:
A safer way to stage this change would be to do a small diff just removing these `__caffe2` registrations, with a heads up to mobile team (cc'ing @linbinyu who added one of these, and @iseeyuan reviewer), wait a little bit, and then land the rest of the changes.

bdhirsh commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -121,7 +121,6 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::embedding_bag_4bit(__torch__.torch.classes.quantized.EmbeddingPackedParamsBase weight, Tensor indices, Tensor? offsets=None, bool scale_grad_by_freq=False, int mode=0, bool pruned_weights=False, Tensor? per_sample_weights=None, Tensor? compressed_indices_mapping=None, bool include_last_offset=False) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::embedding_byte(__torch__.torch.classes.quantized.EmbeddingPackedParamsBase weight, Tensor indices, bool pruned_weights=False) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::celu(Tensor self, float output_scale, int output_zero_point, Scalar alpha=1) -> Tensor"));
-  m.def(TORCH_SELECTIVE_SCHEMA("quantized::hardswish(Tensor input, float output_scale, int output_zero_point) -> Tensor"));
```
random duplicate def found in source :)

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -132,6 +132,12 @@
+    ("__caffe2::BoxWithNMSLimit", datetime.date(2020, 11, 15)),
+    ("__caffe2::BBoxTransform", datetime.date(2020, 11, 15)),
+    ("__caffe2::GenerateProposals", datetime.date(2020, 11, 15)),
+    ("__caffe2::RoIAlignRotated", datetime.date(2020, 11, 15)),
```
just calling out the BC-breakingness again here. each of these ops is registered twice under both the `_caffe` and `__caffe2` namespaces. I couldn't find any usages of the latter version of these 5 ops from a grep search, and I think it makes more sense to hardcode the `_caffe2` namespace directly into the macro, so I opted to remove these 5 ops here. If there's a reason that you think they should still be around, let me know.

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor_test.cpp:
```diff
@@ -68,41 +69,21 @@ TEST(OperatorRegistrationTest_FunctorBasedKernel, givenKernel_whenRegistered_the
   expectCallsIncrement(DispatchKey::CPU);
 }
 
-TEST(OperatorRegistrationTest_FunctorBasedKernel, givenKernel_whenRegistrationRunsOutOfScope_thenCannotBeCalledAnymore) {
```
I ended up deleting this test, since I couldn't figure out a good way to test registration of the same op with different scopes using the legacy API- doing so requires multiple RegisterOperators() instances, which would result in a duplicate def() call for that op.

I rewrote the corresponding test in the lambda/function version of this test file using the new registration API. But I can't do that for this test, because as far as I can tell, the new registration API doesn't allow functors (since the same can be accomplished using a lambda closure). I guess I could have changed this test to use a lambda closure, but that doesn't seem in the spirit of what the other tests in this file are testing.

ezyang replied:
Yeah, deleting this test seems fine. It's testing that duplicate schema registrations work, but we're purposely breaking that, so no need for the test.

bdhirsh commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -46,50 +46,6 @@ struct MockKernel final : OperatorKernel {
   bool* called_;
 };
 
-TEST(OperatorRegistrationTest, whenRegisteringSameSchemaWithAliasAnalysisAfterRegisteringWithoutAliasAnalysis_thenCanBeCalled) {
```
In general, a lot of the tests in this file seemed to test behavior that relies on the fact that you can call RegisterOperators() multiple times for the same op. Since doing so is now an error (doing so causes a duplicate def), I ended up deleting most of those tests.

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "[WIP] make duplicate def an error, fix any resulting issues with the legacy registration API"

Make duplicate definitions in the Dispatcher registration API an error.

Context: The dispatcher has a new API as well as a more legacy API. One advantage of the new API is that it only requires you to specify the full schema of the operator that you're registering in one place (where it is "defined"). All other registrations to different kernels only need the op name and the corresponding kernel. Logically, we'd like to enforce that a user only def()'s an operator once, as this lets us clean up a lot of the AliasAnalysis logic used to deal with duplicate definitions.

The legacy API makes it relatively easier to def() an operator multiple times, and we actually do so in several places in source. Namely:


1. C10_EXPORT_CAFFE2_OP_TO_C10 macros. These macros are used for caffe2 ops, and the macro they use for registration with the dispatcher uses the legacy API in a way that results in duplicate def() calls.

2. Tests. We have lots of tests of the legacy registration API. I tried to clean them up in a sane way- a combination of directly fixing, migrating to the new API, or deleting the tests.

**What this PR does**
1. Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This also required creating two new macros (TORCH_LIBRARY_FRAGMENT_UNIQUE and TORCH_LIBRARY_IMPL_UNIQUE). They are identical to their counterparts, but avoids any variable name collisions when you use each macro multiple times with the same namespace/dispatch key in the same translation unit. This is the case for the caffe2 macros.

2. Updated tests

Only tests using the legacy registration API caused failures. I took all the failed tests and (for the most part) did the following):

(a) Fixed it, if possible, if the test semantically tested the same thing (still using the legacy API)
(b) migrated it to use the new registration API (I did this in tests for which I couldn't figure out a way to get the legacy API to correctly express the test's intent without causing a duplicate def()).
(c) Deleted it (I only did this for tests in op_registration_test. Mostly tests that seemed to be testing the semantic behavior of using duplicate def()'s).

3. Renaming `TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY` to `TORCH_LIBRARY_FRAGMENT`.

Doing this because I'm also planning on migrating custom fb ops to use the new API. Some will require the above macro, since many ops in different files are dumped into the same namespace.

4. Deleting caffe2 operators under the namespace `__caffe2` (NOTE: BC-Breaking).

I did this for two reasons:
(a) I don't see them used anywhere in fbcode, from a grep search. This applies to 5 ops. Each is registered twice (the same operator), under both the `_caffe2` and `__caffe2` namespaces.
(b) Doing so allowed me to hardcode the `_caffe2` namespace into the EXPORT... macro, which I think is preferable to updating all call sites of the macro to pass in a custom namespace. This macro is only meant to be used for caffe2 ops.


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "make duplicate def an error, fix any resulting issues with the legacy registration API"

Make duplicate definitions in the Dispatcher registration API an error.

Context: The dispatcher has a new API as well as a more legacy API. One advantage of the new API is that it only requires you to specify the full schema of the operator that you're registering in one place (where it is "defined"). All other registrations to different kernels only need the op name and the corresponding kernel. Logically, we'd like to enforce that a user only def()'s an operator once, as this lets us clean up a lot of the AliasAnalysis logic used to deal with duplicate definitions.

The legacy API makes it relatively easier to def() an operator multiple times, and we actually do so in several places in source. Namely:


1. C10_EXPORT_CAFFE2_OP_TO_C10 macros. These macros are used for caffe2 ops, and the macro they use for registration with the dispatcher uses the legacy API in a way that results in duplicate def() calls.

2. Tests. We have lots of tests of the legacy registration API. I tried to clean them up in a sane way- a combination of directly fixing, migrating to the new API, or deleting the tests.

**What this PR does**
1. Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This also required creating two new macros (TORCH_LIBRARY_FRAGMENT_UNIQUE and TORCH_LIBRARY_IMPL_UNIQUE). They are identical to their counterparts, but avoids any variable name collisions when you use each macro multiple times with the same namespace/dispatch key in the same translation unit. This is the case for the caffe2 macros.

2. Updated tests

Only tests using the legacy registration API caused failures. I took all the failed tests and (for the most part) did the following):

(a) Fixed it, if possible, if the test semantically tested the same thing (still using the legacy API)
(b) migrated it to use the new registration API (I did this in tests for which I couldn't figure out a way to get the legacy API to correctly express the test's intent without causing a duplicate def()).
(c) Deleted it (I only did this for tests in op_registration_test. Mostly tests that seemed to be testing the semantic behavior of using duplicate def()'s).

3. Renaming `TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY` to `TORCH_LIBRARY_FRAGMENT`.

Doing this because I'm also planning on migrating custom fb ops to use the new API. Some will require the above macro, since many ops in different files are dumped into the same namespace.

4. Deleting caffe2 operators under the namespace `__caffe2` (NOTE: BC-Breaking).

I did this for two reasons:
(a) I don't see them used anywhere in fbcode, from a grep search. This applies to 5 ops. Each is registered twice (the same operator), under both the `_caffe2` and `__caffe2` namespaces.
(b) Doing so allowed me to hardcode the `_caffe2` namespace into the EXPORT... macro, which I think is preferable to updating all call sites of the macro to pass in a custom namespace. This macro is only meant to be used for caffe2 ops.


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

bdhirsh committed:
Update on "make duplicate def an error, fix any resulting issues with the legacy registration API"

Make duplicate definitions in the Dispatcher registration API an error.

Context: The dispatcher has a new API as well as a more legacy API. One advantage of the new API is that it only requires you to specify the full schema of the operator that you're registering in one place (where it is "defined"). All other registrations to different kernels only need the op name and the corresponding kernel. Logically, we'd like to enforce that a user only def()'s an operator once, as this lets us clean up a lot of the AliasAnalysis logic used to deal with duplicate definitions.

The legacy API makes it relatively easier to def() an operator multiple times, and we actually do so in several places in source. Namely:


1. C10_EXPORT_CAFFE2_OP_TO_C10 macros. These macros are used for caffe2 ops, and the macro they use for registration with the dispatcher uses the legacy API in a way that results in duplicate def() calls.

2. Tests. We have lots of tests of the legacy registration API. I tried to clean them up in a sane way- a combination of directly fixing, migrating to the new API, or deleting the tests.

**What this PR does**
1. Update the C10_EXPORT_CAFFE2_OP_TO_C10 macros to use the new dispatcher registration API

This also required creating two new macros (TORCH_LIBRARY_FRAGMENT_UNIQUE and TORCH_LIBRARY_IMPL_UNIQUE). They are identical to their counterparts, but avoids any variable name collisions when you use each macro multiple times with the same namespace/dispatch key in the same translation unit. This is the case for the caffe2 macros.

2. Updated tests

Only tests using the legacy registration API caused failures. I took all the failed tests and (for the most part) did the following):

(a) Fixed it, if possible, if the test semantically tested the same thing (still using the legacy API)
(b) migrated it to use the new registration API (I did this in tests for which I couldn't figure out a way to get the legacy API to correctly express the test's intent without causing a duplicate def()).
(c) Deleted it (I only did this for tests in op_registration_test. Mostly tests that seemed to be testing the semantic behavior of using duplicate def()'s).

3. Renaming `TORCH_LIBRARY_FRAGMENT_THIS_API_IS_FOR_PER_OP_REGISTRATION_ONLY` to `TORCH_LIBRARY_FRAGMENT`.

Doing this because I'm also planning on migrating custom fb ops to use the new API. Some will require the above macro, since many ops in different files are dumped into the same namespace.

4. Deleting caffe2 operators under the namespace `__caffe2` (NOTE: BC-Breaking).

I did this for two reasons:
(a) I don't see them used anywhere in fbcode, from a grep search. This applies to 5 ops. Each is registered twice (the same operator), under both the `_caffe2` and `__caffe2` namespaces.
(b) Doing so allowed me to hardcode the `_caffe2` namespace into the EXPORT... macro, which I think is preferable to updating all call sites of the macro to pass in a custom namespace. This macro is only meant to be used for caffe2 ops.


Differential Revision: [D24567380](https://our.internmc.facebook.com/intern/diff/D24567380)

[ghstack-poisoned]

ezyang commented on torch/library.h:
```diff
@@ -664,6 +667,24 @@ class TorchLibraryInit final {
+///       of each static variable that it creates
+/// It effectively lets you define multiple TORCH_LIBRARY_FRAGMENT blocks
+/// for the same namespace within the same translation unit, avoiding naming collisions.
+#define TORCH_LIBRARY_FRAGMENT_UNIQUE(ns, m, uid) \
```
I remember suggesting in our 1:1 that you might be able to use `C10_ANONYMOUS_VARIABLE` to avoid the need for `UNIQUE` here. Did you end up trying this / did it not work for some reason? In fact, the old macros seem to make use of `__COUNTER__` directly, which suggests that it is supported on all platforms we care about.

bdhirsh replied:
The main problem I have with those (`C10_ANONYMOUS_VARIABLE`/`__COUNTER__`) is that the `TORCH_LIBRARY..` macro's need to re-use whatever variable is created- that doesn't fit with the way that `__COUNTER__` works, which is that it's incremented on every use.

It miiiight be possible to fix that with some extra macro logic (like in [this example](https://stackoverflow.com/questions/4719568/reusable-preprocessor-counter), but it would be pretty ugly (I think I'd need to tailor a new "reuse" macro for each existing library macro.

Unless there's an easier way- then I don't think the ugliness there is better than having to change the regex that you mentioned above.

bdhirsh replied:
The regex values that are used in operator static analysis are [here](https://github.com/pytorch/pytorch/blob/4e6f2440d84a77bd251b7d7e573b0b33e4f3fe5c/tools/code_analyzer/run_analyzer.sh#L23).

I took a look at them, and they shouldn't require any changes to match on the new variable names that I introduced in TORCH_LIBRARY<FRAGMENT|IMPL>_UNIQUE:
```
^.*TORCH_LIBRARY_FRAGMENT_init_([^(]+)(\(.*)?$
^.*TORCH_LIBRARY_IMPL_init_([^(]+)_([^_]+)(\(.*)?$
```
Between the `_init_` and the left paren of the function, both bits of regex match on arbitrary characters (up until the left paren), so adding a counter to the end of the variable name shouldn't affect which ones do/don't get matched.

I wasn't aware of this operator static analysis though- @ljk53 is there anything to watch out for, given that this PR (and the corresponding changes in fbcode) will be updating the caffe2 ops + fbcode ops to use these macros? Those ops will now all become a part of the analysis.

ezyang replied:
> Unless there's an easier way- then I don't think the ugliness there is better than having to change the regex that you mentioned above.

The payoff for having the double macro structure is you get to avoid having both UNIQUE and non-UNIQUE public facing API. The test is: internal complication is not as bad, as long as it is encapsulated, but if the complication leaks out (as it does for UNIQUE), we should try to avoid it.

ezyang replied:
> Between the _init_ and the left paren of the function, both bits of regex match on arbitrary characters (up until the left paren), so adding a counter to the end of the variable name shouldn't affect which ones do/don't get matched.

I don't think this analysis is enough. The regex is not only used to match; the groups matter too (as seen in the use of `return pattern.pattern->sub("\\1", functionName) + "::";`. In particular, if there is extra goop after the namespace, it will be picked up too and you will infer the wrong namespace here.

By the way, this is another good reason to not have two variants of the macro. I think the reason why the existing tests do not catch this is they do not exercise the Caffe2 exported operator codepath (since you're not really supposed to put Caffe2 operators in selective build). If you applied the anonymous variable treatment universally, I bet the selective build tests would have started failing.

bdhirsh replied:
You're completely right, thanks.

> The payoff for having the double macro structure is you get to avoid having both UNIQUE and non-UNIQUE public facing API

I guess I thought that we could still consider those macros "private" if they were only used within pytorch source code (which is probably wrong). Also, it doesn't actually look that hard to add in the extra macro logic (I'll update the PR shortly with that). Tentatively, I'm updating the existing macros to call out to (truly private) versions of each macro that take in a unique identifier.

> In particular, if there is extra goop after the namespace, it will be picked up too and you will infer the wrong namespace here

Yeah, I didn't look closely enough at this. I just looked at it some more and finally figured out how to run the op analyzer and see / play with the results. With the macro changes that I'm planning on adding, you're right that I think I'll need to update the regex to make sure it correctly pulls out the namespace. As an aside, I think that the existing `TORCH_LIBRARY_FRAGMENT` macro isn't getting its namespace read properly either, since there's a stray `_k` appended to the variable (that gets included in the namespace in the regex) [here](https://github.com/pytorch/pytorch/blob/4e6f2440d84a77bd251b7d7e573b0b33e4f3fe5c/torch/library.h#L659). I'm not sure what/if that actually breaks anything, but it would only affect the ops where that macro is used, [here](https://github.com/pytorch/pytorch/blob/4e6f2440d84a77bd251b7d7e573b0b33e4f3fe5c/aten/src/ATen/native/RNN.cpp#L1862)

ezyang replied:
> I guess I thought that we could still consider those macros "private" if they were only used within pytorch source code (which is probably wrong)

You're right that it's a little wibbly wobbly. One relevant abstraction barrier here is things that are header internal, versus things that get used outside of the header. That's the sense of privacy I'm going for here, though it's definitely a convention.

>  As an aside, I think that the existing TORCH_LIBRARY_FRAGMENT macro isn't getting its namespace read properly either, since there's a stray _k appended to the variable (that gets included in the namespace in the regex) here.

Oops, that sounds like it indeed would be problematic. Actually, the k is probably due to copy paste from `TORCH_LIBRARY_IMPL` and could just be deleted entirely. It's probably not hitting tests because all the registrations in PyTorch proper are just using `TORCH_LIBRARY`; we used to generate fragment registrations for the actual mobile deploy but (1) we never run code analysis on that, and (2) we got rid of that code.

ezyang commented:
> Deleting caffe2 operators under the namespace __caffe2 (NOTE: BC-Breaking).

One other check you can do on this is find the PR that added them and check if they reference anything specific.

ezyang commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -920,8 +611,11 @@ TEST(OperatorRegistrationTest, whenRegisterWithXLAKernelAndCatchAll_AutogradXLAI
 
-  auto registrar1 = c10::RegisterOperators().op("_test::dummy(Tensor dummy) -> ()", c10::RegisterOperators::options()
-    .kernel<decltype(autograd_kernel), &autograd_kernel>(DispatchKey::XLA));
+  registrar = c10::RegisterOperators(); // destruct
```
nit: probably clearer to just scope the first registration

ezyang reviewed (commented):
This is good work. High level commentary:

- Let's do the `__caffe2` removal in a separate diff, and CC mobile team on it
- If we can avoid the `_UNIQUE` versions of the macros entirely, that would be better. If you haven't tried the anonymous variable macro yet, it's worth at least a try before we move forward with this patch as is.

Also, I think I mentioned that changing the name of the static registration affects the operator static analysis (@ljk53).  This is done by:

```
cl::list<RegexOpt, bool, RegexOptParser> TorchLibraryInitPattern(
    "torch_library_init_pattern",
    cl::desc("Regular expression used to identify TorchLibraryInit symbols "
             "that are generated by `TORCH_LIBRARY` macro. The first capturing "
             "group is used to extract namespace string. "
             "Example: -torch_library_init_pattern "
             "'^.*TORCH_LIBRARY_init_([^(]+)(\\(.*)?$'"),
    cl::ZeroOrMore);

...

  static std::string inferContextualNamespace(Instruction* I) {
    auto functionName = _demangle(I->getFunction()->getName());
    for (auto& pattern : TorchLibraryInitPattern) {
      if (!pattern.pattern->match(functionName)) {
        continue; 
      }
      if (Verbose) {
        std::cerr << "[DEBUG][REG][INIT_FUNC] " << functionName << std::endl;
      } 
      return pattern.pattern->sub("\\1", functionName) + "::";
    } 
    return {};
  }   
```

With the new function naming convention to allow multiple occurrences, contextual namespace inference is broken. Fortunately, `__COUNTER__` is guaranteed to give numbers, so it's easy enough to adjust the regex (however, this will be easier if all functions have a uniform naming convention; thus circling back to my comment about trying to use anonymous variables for the names unconditionally).

bdhirsh commented:
> * Let's do the `__caffe2` removal in a separate diff, and CC mobile team on it

In general, I think I should try to split my PR up into smaller pieces (including within fbcode), so I'll start on that.

I git blame'd the `__caffe2` stuff (thanks for the callout, probably should have done that to start out) and from [the PR here](https://github.com/pytorch/pytorch/commit/b704f30189a6a9f6551b0554ddc82ae5b9241e2c) it sounds like that separate namespace is required for the lite interpreter. @linbinyu is that right that I should keep them in? It'll just require a bit more refactoring.

> * If we can avoid the `_UNIQUE` versions of the macros entirely, that would be better. If you haven't tried the anonymous variable macro yet, it's worth at least a try before we move forward with this patch as is.

See my comment in the file- I couldn't get it to work easily because of the fact that each macro needs to define + use the same variable. __COUNTER__ doesn't have a clean way of letting you re-use its value.

> Also, I think I mentioned that changing the name of the static registration affects the operator static analysis (@ljk53). 

Also see my comment in the file. I ended up modifying the caffe2 macro slightly to pass in __COUNTER__ instead of the OperatorName as a unique identifier (if you think that's not worth the changes, let me know and I can revert). But either way, I think that the existing regex should still match the new variable names.

bdhirsh commented:
I cleaned this up and separated it out into several smaller diffs here (sadly comments will not transfer over):
https://github.com/pytorch/pytorch/pull/47318

I also moved all of the fb-specific changes to a separate diff (check the stack in phabricator)

ezyang commented:
> See my comment in the file- I couldn't get it to work easily because of the fact that each macro needs to define + use the same variable. __COUNTER__ doesn't have a clean way of letting you re-use its value.

See https://stackoverflow.com/a/13605588/23845

ezyang commented:
> Also see my comment in the file. I ended up modifying the caffe2 macro slightly to pass in __COUNTER__ instead of the OperatorName as a unique identifier (if you think that's not worth the changes, let me know and I can revert). But either way, I think that the existing regex should still match the new variable names.

Alright. I mentioned this because I didn't see any remark of this in the PR.
